### ISSUE TITLE
Need a bestvideo+bestcompatibleaudio format selector

### ISSUE THREAD

**Issue by Daemonster** [rg3/youtube-dl#10226]
### I have
- [x] **Verified** and **I assure** that I'm running youtube-dl **2016.08.01**
- [x] At least skimmed through [README](https://github.com/rg3/youtube-dl/blob/master/README.md) and **most notably** [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### The purpose of this *issue*
- [ ] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [x] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

### Description
This request follows [this question](https://github.com/rg3/youtube-dl/issues/10221).

I need a format selector that is something like **`-f bestvideo+bestcompatibleaudio`**, where `'bestcompatibleaudio'` is the best audio file which is compatible with the desired video file (so that the files don't need to be muxed into an mkv.) Or, if you all find this cool, maybe even `'worstcompatibleaudio'`

_Until it is implemented, anyone who is in dire need of this feature can use this [script](https://gist.github.com/Daemonster/4c94372a85cd4b5ea86df06e102a19e7)._

**Comment by dstftw** [comment]
No, compatible-shortcuts is not a way to do. Compatibility is the characteristic of the merging operation not a particular format. Thus instead of introducing another n shortcuts an alternative *compatible-merge-operator* should be introduced.

**Comment by Daemonster** [comment]
> Compatibility is the characteristic of the merging operation

I guess you're referring to the compatibility of codecs. Yes, that's not wrong. I already see a TODO [here](https://github.com/rg3/youtube-dl/blob/master/youtube_dl/YoutubeDL.py#L1651). Maybe a `codec-compatible-formats-selection` feature would be great to have.

**Comment by dstftw** [comment]
I'm not referring to codecs. `codec-compatible-formats-selection` will make it impossible to mix *standard-merge* and *compatible-merge* in single format selection expression.

**Comment by Daemonster** [comment]
I'm too lazy to scour the source. Mind sharing a link to the code you're referring to?

**Comment by dstftw** [comment]
I'm not referring to any code. It's not implemented.

**Comment by Daemonster** [comment]
Oh, my bad. Anyway, the point is, I am not talking _specifically_ about the implementation of such a format selector. Rather, I am suggesting that we need a feature (maybe a new interface option) that lets us make `youtube-dl` automatically choose (if possible) the audio and video formats, based on their codecs, such that they are compatible and don't need to be muxed into matroska.

**Comment by weedy** [comment]
You can already "block" mkv. You're just not reading the docs.
From my mplayer config:
```
ytdl-format=298+bestaudio[ext=m4a]/137+bestaudio[ext=m4a]/bestvideo[ext=mp4][height<?1080]+bestaudio[ext=m4a]/best
```

**Comment by Daemonster** [comment]
Wow! Either I am horrible at explaining things or you are not that great at understanding things. I _read_ the docs and I am aware of that. My focus is NOT on _blocking mkv_, rather, it is on _ensuring_ that video and audio files of compatible extensions are downloaded - for example, an m4a DASH video along with an mp4 DASH audio or a WebM DASH video along with a WebM DASH audio, NOT an mp4 DASH video along with a WebM DASH audio along, even if it is the best available audio format (which is usually the case with YouTube videos.) Did you at least read the title of the issue before commenting?

This is what I am talking about, in pseudo-code:

    bestvideo = getbestvideo()
    bve = extension(bestvideo)
    bestcompatibleaudio = Nothing
    for a in sort_in_descending_order_of_quality(audiofiles):
        if iscompatible(extension(a), bve):
            bestcompatibleaudio = a
            break
    selected_formats = (bestvideo, bestcompatibleaudio or getbestaudio())

**Comment by yan12125** [comment]
@dstftw Did you mean something like this: ```bestvideo$bestaudio``` (forgive me for the naive selection of ```$```)? How to know whether it's the best video with the best compatible one among audio formats, or best audio with the best compatible one among video formats?

**Comment by dstftw** [comment]
>Did you mean something like this: ```bestvideo$bestaudio```

Yes, that's the general idea.

>How to know whether it's the best video with the best compatible one among audio formats, or best audio with the best compatible one among video formats?

For instance, by introducing another *audio-plus-compatible-video-merge* operator.

**Comment by yan12125** [comment]
Sounds good.

**Comment by Daemonster** [comment]
@dstftw,
> For instance, by introducing another audio-plus-compatible-video-merge operator.

Actually, `video-plus-compatible-audio-merge` is what I had in mind when I opened this issue, rather than `audio-plus-compatible-video-merge`. Or maybe we can have both ways?
BTW, the introduction of an operator was a smart idea.

@yan12125,
May I suggest the usage of `*` rather than `$`?
And thanks for listening and adding the label!

**Comment by dstftw** [comment]
Obviously we should have both since both scenarios are viable.

**Comment by videonerd** [comment]
Any plans to realize this awesome feature directly in the program?

**Comment by endquote** [comment]
Also wishing for this. The merging of incompatible formats into mkv is great, but I'm finding that those files playback incorrectly in Chrome (but work fine in VLC).

Of course this is not ytdl's fault, but still.
